File: questions/159. Longest Substring with At Most Two Distinct Characters/longest.py
class Solution:
    def lengthOfLongestSubstringTwoDistinct(self, s: str) -> int:
        l, res, left, right = len(s), 0, 0, 0
        dic = defaultdict(int)
        while right < l:
            dic[s[right]] += 1
            if len(dic) > 2:
                while len(dic) > 2:
                    dic[s[left]] -= 1
                    if dic[s[left]] == 0:
                        del dic[s[left]]
                    left += 1
            res = max(res, right - left + 1)
            right += 1
        return res
                    
        

File: questions/159. Longest Substring with At Most Two Distinct Characters/longest.cpp
class Solution {
public:
    int lengthOfLongestSubstringTwoDistinct(string s) {
        int l = s.size(), left = 0, right = 0, res = 0;
        unordered_map<char, int> m;
        while(right < l){
            m[s[right]]++;
            if(m.size() > 2){
                while(m.size() > 2){
                    m[s[left]]--;
                    if(!m[s[left]]){
                        m.erase(s[left]);
                    }
                    left++;
                }    
            }
            res = max(res, right - left + 1);
            right++;
        }
        return res;
    }
};
